contracts/staking/src/contract.rs

```diff
@@ -614,27 +618,31 @@ pub fn redeem_rewards(
     mut deps: DepsMut,
     _info: MessageInfo,
     env: Env,
+    validators: Option<Vec<Addr>>,
 ) -> Result<Response, ContractError> {
-    check_slashing(&mut deps, &env)?;
+    let all_delegations = deps
+        .querier
+        .query_all_delegations(env.contract.address.clone())?;
+    let all_validators = deps.querier.query_all_validators()?;
+
+    check_slashing(&mut deps, &env, all_delegations.as_slice())?;
     let state = STATE.load(deps.storage)?;
+
+    let validators = validators.unwrap_or(state.validators);
     let operation_controls = OPERATION_CONTROLS.load(deps.storage)?;
     if operation_controls.redeem_rewards_paused {
         return Err(ContractError::OperationPaused("redeem_rewards".to_string()));
     }
 
+    let mut attrs = vec![];
     let mut messages = vec![];
     let mut failed_vals: Vec<String> = vec![];
-    for val_addr in state.validators {
+    for val_addr in validators {
```

Comment:
Doesnt it make sense to just loop through all_delegation validators?

Reply:
No dude it might timeout. If we have a large set of validators. We ideally go thru all the validators in the validator list but if not then we just go thru the validator list we pass in. This was just of the optimizations.